torch/csrc/dynamo/guards.cpp

```diff
@@ -622,5 +987,82 @@ PyObject* torch_c_dynamo_guards_init() {
     return nullptr;
   }
 
+  auto py_m = py::handle(m).cast<py::module>();
+  py::class_<GuardDebugInfo, std::unique_ptr<GuardDebugInfo>>(
+      py_m, "GuardDebugInfo")
+      .def(py::init<int, std::string>())
+      .def_readonly("num_guards_executed", &GuardDebugInfo::num_guards_executed)
+      .def_readonly("failure_reason", &GuardDebugInfo::failure_reason)
+      .def("__repr__", &GuardDebugInfo::repr);
+
+  // Leaf Guards
+  py::class_<LeafGuard, std::unique_ptr<LeafGuard>>(py_m, "LeafGuard");
+  py::class_<PythonLambdaGuard, LeafGuard, std::unique_ptr<PythonLambdaGuard>>(
+      py_m, "PythonLambdaGuard")
+      .def(py::init<py::function, py::function>())
+      .def("__call__", &PythonLambdaGuard::check);
+
+  // Guard Accessors - These are present so that we can iterate over the
+  // GuardManager hierarchy. We intentionally do not provide even an init
+  // function on these, because these should be constructed from within C++.
+  py::class_<GuardAccessor, std::unique_ptr<GuardAccessor>>(
+      py_m, "GuardAccessor");
+  py::class_<
+      AttrGuardAccessor,
+      GuardAccessor,
+      std::unique_ptr<AttrGuardAccessor>>(py_m, "AttrGuardAccessor");
+  py::class_<
+      ItemGuardAccessor,
+      GuardAccessor,
+      std::unique_ptr<ItemGuardAccessor>>(py_m, "ItemGuardAccessor");
+  py::class_<
+      PythonLambdaGuardAccessor,
+      GuardAccessor,
+      std::unique_ptr<PythonLambdaGuardAccessor>>(
+      py_m, "PythonLambdaGuardAccessor");
+
+  // Guard Manager
+  py::class_<GuardManager, std::unique_ptr<GuardManager>>(py_m, "GuardManager")
+      .def(py::init<>())
+      .def("check", &GuardManager::check)
```

Comment:
Another optimization I would like to see here is calling guards directly via C++ rather than round-tripping through Python.  If we do this, then this check method doesn't need to be bound here. 

The goal here is the pybind should only be used for guard construction, not guard checking.

Reply:
There is fast path now - `check_pybind`. We will use this when we stick the GuardManager in cache entry. For debugging purpose, I still have a `check` function exposed to Python.